### PR TITLE
Support TTBR1 as well as TTBR0

### LINKED ISSUE #14: Support setting `TTBR1_EL1` too
`IdMap::activate` currently assumes that only `TTBR0_EL1` is used. We should support other configurations too.

### PR DESCRIPTION
**Pull request by qwandor**
Fixes #14.

### CODE REVIEW

File: src/paging.rs [hunk 1]
```diff
@@ -25,6 +25,17 @@ pub const PAGE_SIZE: usize = 1 << PAGE_SHIFT;
 /// page size.
 pub const BITS_PER_LEVEL: usize = PAGE_SHIFT - 3;
 
+/// Which TTBR register to use for a page table.
```
Comment (AlanStokes) [inline src/paging.rs:28]: Maybe it's easiest to use the register names, but you could alternatively be slightly more abstract and  refer to the lower or upper VA range; that's the ARM ARM terminology.

File: src/linearmap.rs [hunk 2]
```diff
@@ -318,10 +373,29 @@ mod tests {
...
+    #[test]
+    fn virtual_address_range_ttbr1() {
+        const GIB_512: isize = 512 * 1024 * 1024 * 1024;
+        // Map the 512 GiB region at the top of virtual address space to the bottom of physical
+        // address space.
+        let translation = LinearTranslation::new(GIB_512);
+
+        // The first page in the region covered by TTBR1.
+        assert_eq!(
+            translation.virtual_to_physical(VirtualAddress(0xffffff8000000000)),
```
Comment (AlanStokes) [inline src/linearmap.rs:385]: Could we add underscores to make these more readable - 0xffff_ff80_0000_0000? Or compute it some other way, so it's obvious? -GIB_512, perhaps.

File: src/linearmap.rs [hunk 3]
```diff
@@ -318,10 +373,29 @@ mod tests {
         )
     }
 
+    #[test]
+    fn virtual_address_range_ttbr1() {
+        const GIB_512: isize = 512 * 1024 * 1024 * 1024;
```
Comment (AlanStokes) [inline src/linearmap.rs:378]: Extract to top level rather than repeating?

File: src/paging.rs (hunk 1 above)
Comment (qwandor) [reply src/paging.rs]: Done.

File: src/linearmap.rs (hunk 3 above)
Comment (qwandor) [reply src/linearmap.rs]: Done.

File: src/linearmap.rs (hunk 2 above)
Comment (qwandor) [reply src/linearmap.rs]: Done.

**Review by ardbiesheuvel — COMMENTED** [review]
The linear map can by definition only be used with TTBR0 so this patch seems wrong or unnecessary at least?

**Comment by qwandor** [comment]
Uh, why can't the linear map be used with `TTBR1`? It seems reasonable to use `TTBR1` to map some or all of physical address space to the upper virtual address space at a fixed offset.

**Comment by ardbiesheuvel** [comment]
> Uh, why can't the linear map be used with `TTBR1`? It seems reasonable to use `TTBR1` to map some or all of physical address space to the upper virtual address space at a fixed offset.

Apologies, you're quite right - I got confused and read 'identity map' instead of 'linear map'. Ignore me ...

File: src/paging.rs [hunk 4]
```diff
@@ -25,15 +25,15 @@ pub const PAGE_SIZE: usize = 1 << PAGE_SHIFT;
...
-    Ttbr0,
-    /// `TTBR1`, meaning that the page table covers the top of the virtual address space (ending at
-    /// address 0xffffffffffffffff).
-    Ttbr1,
+pub enum VaRange {
+    /// The page table covers the bottom of the virtual address space (starting at address 0), so
+    /// will be used with `TTBR0`.
+    Lower,
+    /// The page table covers the top of the virtual address space (ending at address
+    /// 0xffffffffffffffff), so will be used with `TTBR1`.
```
Comment (AlanStokes) [inline src/paging.rs:35]: nit: 0xffff_ffff_ffff_ffff

**Review by AlanStokes — APPROVED** [review]

**Review by ardbiesheuvel — APPROVED** [review]

File: src/paging.rs (hunk 4 above)
Comment (qwandor) [reply src/paging.rs]: Done.

**Pull request merged**